scripts/gen_mmu.py

```diff
@@ -7,75 +7,1001 @@
+        return((value >> 21) & 0x1FF)
+
+    # return the page table number for the given address
+    def get_pte_number(self, value):
+        return((value >> 12) & 0x1FF)
+
+    def get_total_number_of_pt(self):
+        return(len(pd_tables_list))
+
+    def get_number_of_pd(self):
+        return(len(list(set(map(lambda x: x[0], pd_tables_list)))))
+
+    def get_pdpte_list(self):
+        return(list(set(map(lambda x: x[0], pd_tables_list))))
+
+    # the return value will have the page address and it is assumed to be a 4096
+    # boundary.hence the output of this API will be a 20bit address of the page
+    # table
+    def address_of_page_table(self, pdpte, page_table_number):
+        global pd_start_addr
+
+        # first page given to page directory pointer
+        # and 2nd page till 5th page are used for storing the page directories.
+
+        # set the max pdpte used. this tells how many pd are needed after
+        # that we start keeping the pt
+        PT_start_addr = self.get_number_of_pd() * 4096 +\
+            pd_start_addr + 4096
+        return (PT_start_addr +
+                (pd_tables_list.index([pdpte, page_table_number]) *
+                 4096) >> 12)
+
+    #   union x86_mmu_pae_pde {
+    # 	u64_t  value;
+    # 	struct {
+    # 		u64_t p:1;
+    # 		u64_t rw:1;
+    # 		u64_t us:1;
+    # 		u64_t pwt:1;
+    # 		u64_t pcd:1;
+    # 		u64_t a:1;
+    # 		u64_t ignored1:1;
+    # 		u64_t ps:1;
+    # 		u64_t ignored2:4;
+    # 		u64_t page_table:20;
+    #           u64_t igonred3:29;
+    #           u64_t xd:1;
+    # 	};
+    # };
+
+    def get_binary_pde_value(self, pdpte, value):
+        perms = value.page_entries_info[0].permissions
+
+        present = 1 << 0
+        read_write = check_bits(perms, [1, 29]) << 1
+        user_mode = check_bits(perms, [2, 28]) << 2
+
+        pwt = 0 << 3
+        pcd = 0 << 4
+        a = 0 << 5  # this is a read only field
+        ps = 0 << 7  # set to make sure that the phy page is 4KB
+        page_table = self.address_of_page_table(pdpte, value.pde_index) << 12
+        xd = check_bits(perms, [62, 63]) << 63
+        return (present |
+                read_write |
+                user_mode |
+                pwt |
+                pcd |
+                a |
+                ps |
+                page_table |
+                xd)
+
+    # union x86_mmu_pae_pte {
+    # 	u64_t  value;
+    # 	struct {
+    # 		u64_t p:1;
+    # 		u64_t rw:1;
+    # 		u64_t us:1;
+    # 		u64_t pwt:1;
+    # 		u64_t pcd:1;
+    # 		u64_t a:1;
+    # 		u64_t d:1;
+    # 		u64_t pat:1;
+    # 		u64_t g:1;
+    # 		u64_t ignore:3;
+    # 		u64_t page:20;
+    #           u64_t igonred3:29;
+    #           u64_t xd:1;
+    # 	};
+    # };
+    def get_binary_pte_value(self, value, pde, pte, perm_for_pte):
+        present = 1 << 0
+        read_write = ((perm_for_pte >> 1) & 0x1) << 1
+        user_mode = ((perm_for_pte >> 2) & 0x1) << 2
+        pwt = 0 << 3
```

Comment:
Lots of these bits will be set to 0, and many of these shifts will be used and reused throughout these functions to generate PTEs, PDEs, etc.  Might be better to define constants that can be ORed together and get rid of the comment with the C struct definition:

```Python

ret = PTE_PRESENT
ret |= PTE_READ_WRITE if perm_for_pte & PERM_READ_WRITE else 0
ret |= PTE_USER_MODE if perm_for_pte & PERM_USER_MODE else 0
...

return ret
```

There's no need to specify anything if the fields are not set.  Not a huge deal as well.

Reply:
Did a variation of this but still kept the c struct. I hope that it would help understand what is happening in the code.